### PR TITLE
[ENG-1987] indexer shenanigans

### PR DESCRIPTION
**Pull request by aaxelb**
ok so the point of this is to make it so the elastic index can be populated directly from `NormalizedData` instead of `AbstractCreativeWork`, to make way for deleting `ShareObject` and all its kin (including `AbstractCreativeWork`) -- all without breaking the apps we've built that use the index directly (preprints and registries discover pages -- will be updated to use a new search api once that new api exists)

## changes
- docker-compose updates
  - add `indexer` service
  - move common environment variables to `.docker-compose.env`
  - tell `celery` its app
- `share.search` tidy-up
  - cleaner indexer daemon threading, messaging
  - remove unhelpfully redundant code
  - move elastic mappings and fetching logic to IndexSetup classes
    - available IndexSetups added to entry_points in setup.py
    - allow configuring a different IndexSetup for each index
    - currently only one `share_classic` IndexSetup that just wraps
      existing behavior
  - move index creation/deletion to ElasticManager
- ensure useful suids for all datums
  - require suids for pushed data
  - for compatibility, infer suids for data pushed from OSF
  - `populate_osf_suids` command to populate suids for existing OSF data
- add `FormattedMetadataRecord` model
  - each suid has one FMR for each supported metadata format (see the list in `setup.py` under `entry_points => share.metadata_formats`)
  - the elasticsearch source document is now stored as an FMR during `ingest`, and the indexer daemon just pulls from that table (makes re-indexing much smoother)

## todo
- [x] tests for ElasticManager
- [x] tests for IndexSetup(s)
- [x] (better) tests for SearchIndexerDaemon
- [x] `postrend_backcompat` IndexSetup that builds elastic documents from NormalizedData

### CODE REVIEW

**Review by felliott — CHANGES_REQUESTED** [review]
Looking good!  Just a few comments.

File: share/util/osf.py [hunk 1]
```diff
@@ -10,3 +14,33 @@ def osf_sources():
     ).exclude(
         user__username=settings.APPLICATION_USERNAME,
     )
+
+
+OSF_GUID_RE = re.compile(r'^https?://(?:[^.]+.)?osf.io/(?P<guid>[^/]+)/?$')
```
[the remark this answers is not in the document]
Comment (aaxelb) [reply share/util/osf.py:19]: ah whoops, good catch!

[the remark this answers is not in the document]
Comment (aaxelb) [reply share/util/osf.py:19]: yeah that was kinda intentional -- i don't know if "exactly 5" will always be set in stone (do we have a way to tell how much of valid-guid-space is used/left?) and it seemed safe enough to allow any length, as long as it's not a nested path

File: api/normalizeddata/views.py [hunk 2]
```diff
@@ -65,8 +67,14 @@ def create(self, request, *args, **kwargs):
         serializer = self.get_serializer_class()(data=request.data, context={'request': request})
         serializer.is_valid(raise_exception=True)
         with transaction.atomic():
+            data = serializer.validated_data['data']
+            suid = serializer.validated_data.get('suid', None)
+            if not suid:
+                # HACK: try for an osf guid -- may still be None tho
+                suid = guess_osf_guid(MutableGraph.from_jsonld(data))
+
```
[the remark this answers is not in the document]
Comment (aaxelb) [reply api/normalizeddata/views.py:75]: good point, can be pre-computed

**Pull request merged**